### ISSUE TITLE
Start supporting pydantic v2

### ISSUE THREAD

**Issue by Hofer-Julian** [Deltares/Ribasim#382]
pydantic v2 is now out with a lot of nice things, most notably higher speed.

I think we will want to support v2 as fast as possible, but depending on the amount of work, it would be nice to continue supporting v1 as well.
If we don't do that, packaging Ribasim Python on Deltaforge will be hard to do in the near future because of version requirement conflicts.

The migration guide can be found here: https://docs.pydantic.dev/latest/migration/

**Comment by Hofer-Julian** [comment]
Fully moving to pydantic v2 depends on https://github.com/machow/quartodoc/pull/211

**Comment by evetion** [comment]
And on Pandera, which has a Pydantic V2 compatible version coming in October at the earliest (and we'll have to test it first).

**Comment by visr** [comment]
Pandera now supports pydantic v2 as of v0.17: https://github.com/unionai-oss/pandera/releases/tag/v0.17.0
Quartodoc merged support a few days ago to main, so the upcoming release after v0.6.3 should have it.

**Comment by visr** [comment]
Ok our dependencies are ready, now we just need to update.

I tried `pixi add pydantic=2` and that works, only it causes ribasim-python tests to fail.

```
E       pydantic_core._pydantic_core.ValidationError: 5 validation errors for Model
E       outlet
E         Field required [type=missing, input_value={'modelname': 'basic', 'n...='debug', timing=False)}, input_type=dict]
E           For further information visit https://errors.pydantic.dev/2.4/v/missing
E       discrete_control
E         Field required [type=missing, input_value={'modelname': 'basic', 'n...='debug', timing=False)}, input_type=dict]
E           For further information visit https://errors.pydantic.dev/2.4/v/missing
E       pid_control
E         Field required [type=missing, input_value={'modelname': 'basic', 'n...='debug', timing=False)}, input_type=dict]
E           For further information visit https://errors.pydantic.dev/2.4/v/missing
E       user
E         Field required [type=missing, input_value={'modelname': 'basic', 'n...='debug', timing=False)}, input_type=dict]
E           For further information visit https://errors.pydantic.dev/2.4/v/missing
E       solver
E         Field required [type=missing, input_value={'modelname': 'basic', 'n...='debug', timing=False)}, input_type=dict]
E           For further information visit https://errors.pydantic.dev/2.4/v/missing

python\ribasim_testmodels\ribasim_testmodels\basic.py:196: ValidationError
```

I don't really understand why since these are declared as Optional. Probably best to wait until @evetion tackled #630. @evetion I don't know if you plan to upgrade as part of the redesign, but it's good to know that you could start using pydantic v2 features. You could use d73637193011a8a7fc08c1812b9238e1e27f5f54 for the dependency upgrade if needed.

**Comment by Hofer-Julian** [comment]
Optional (or T | None) behaves differently with pydantic 2 (another reason why we should drop the misleading Optional and use the new syntax instead)

**Comment by Hofer-Julian** [comment]
@visr that is the change: https://docs.pydantic.dev/latest/migration/#required-optional-and-nullable-fields

TLDR: just add a default value of `None` and you get the old behavior

**Comment by visr** [comment]
Fixed in #731
